src/main/drivers/compass_ist8310.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <math.h>
+
+#include <platform.h>
+#include "build/debug.h"
+
+#include "common/axis.h"
+#include "common/maths.h"
+
+#include "config/parameter_group.h"
+
+#include "system.h"
+#include "nvic.h"
+#include "gpio.h"
+#include "bus_i2c.h"
+#include "light_led.h"
+
+#include "sensor.h"
+#include "compass.h"
+
+#include "sensors/sensors.h"
+#include "compass_ist8310.h"
+
+//#define DEBUG_MAG_DATA_READY_INTERRUPT
+
+// ist8310, default address 0x1C
+// NAZE Target connections
+// PB12 connected to MAG_DRDY on rev4 hardware
+// PC14 connected to MAG_DRDY on rev5 hardware
+
+/* CTRL_REGA: Control Register 1
+ * Read Write
+ * Default value: 0x0A
+ * 7:4  0   Reserved.
+ * 3:0  DO2-DO0: Operating mode setting
+ *        DO3  |  DO2 |  DO1 |  DO0 |   mode
+ *    ------------------------------------------------------
+ *         0   |   0  |  0   |  0   |   Stand-By mode
+ *         0   |   0  |  0   |  1   |   Single measurement mode
+ *                                       Others: Reserved
+ *
+ * CTRL_REGB: Control Register 2
+ * Read Write
+ * Default value: 0x0B
+ * 7:4  0   Reserved.
+ * 3    DREN : Data ready enable control: 
+ *      0: disable 
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define MAG_ADDRESS 0x0C
+#define MAG_DATA_REGISTER 0x03
+#define MAG_WHOAMI 0x00
+
```

Comment:
Yes ... some CF code is a bit ugly. But it would be better to add high-quality code.
You should take a look at ak drivers ... the code seems to be cleaner than HMC.